xlito.man
.\"	@(#)x11:contrib/clients/xloadimage/xlito.man 1.3 93/07/29 Labtam
.\"	
.TH XLITO 1 "7 Jul 1993"
.SH NAME
xlito (XLoadImageTrailingOptions) \- Append/Delete/Show a Trailing Option 
string in a file.
.SH SYNOPSIS
\fIxlito\fR [option] [string] files ...
.SH DESCRIPTION
\fIxlito\fR (XLoadImageTrailingOptions) is a utility that provides
a file format independent way of marking image files with the appropriate
options to display correctly.  It does this by appending to a file, a string
specified by the user. The string is marked with some magic numbers so that
it can be extracted by a program that knows what to look for. Since almost all
image files have some sort of image dimension information in the file, the
programs that load or manipulate these files generally do not look beyond
the point at which they have read the complete image, so trailing information
is safely be hidden. If appending this information causes trouble with other
utilities, it can simply be deleted.
.PP
Appropriate version of \fIxloadimage\fR (ie. \fIxli 1.00\fR) will recognise
these trailing options at the end of the image files, and will treat the
embedded string as if it were a sequence of command line \fIImage Options\fR.
Any \fIGlobal Options\fR will be ignored, and unlike command line
\fIimage Options,\fR \fITrailing Options\fR are never propagated to other
images.
.PP
Old versions of \fIxloadimage\fR (3.01 or less) can be made forward compatible
by using the \fI-x\fR option to pull the trailing options out of the image files,
and put them on the command line where \fIxloadimage\fR can see them.
.SH OPTIONS
The default behavior is to display the trailing option strings (if any) of
the files on the argument line.  The following options change the behavior of
\fIxlito:\fR
.TP 8
-c \fIoption_string file_name ...\fR
This adds or changes the embedded string to \fIoption_string\fR. 
The string will have to be quoted if it is composed of more than one word.
.TP
-d \fIfile_name ...\fR
Delete any embedded trailing option strings in the given files.
.TP
-x \fIfile_name ...\fR
Process the files and create a command line string suitable for use by
\fIxloadimage.\fR Arguments starting with \fI-\fR are echoed, arguments
not starting with \fI-\fR are treated as files and any trailing options
strings are echoed followed by the file name. The \fIxloadimage\fR option
\fI-name\fR is treated correctly. 
.SH EXAMPLES
If fred.gif has the wrong aspect ratio, then it might need
viewing with the \fIxloadimage\fR options:
.sp
.ti +5
xloadimage -yzoom 130 fred.gif
.PP
This option can then be appended to the file by:
.sp
.ti +5
xlito -c "-yzoom 130" fred.gif
.PP
and from then on some new versions of xloadimage will get the appropriate options
from the image file itself. Old versions of \fIxloadimage\fR can be made to
work by using:
.sp
.ti +5
xloadimage \(gaxlito -x fred.gif\(ga
.PP
This can be made transparent by using a script containing something like:
.sp
.ti +5
xloadimage \(gaxlito -x $\(**\(ga
.PP
The script could be called \fIxli\fR for instance.
.PP
The options can be deleted with:
.sp
.ti +5
xlito -d fred.gif
.SH AUTHOR
.nf
Graeme Gill
Labtam Australia
graeme@labtam.oz.au
.fi
.SH COMPATIBILITY WITH IMAGE FILES
Some image files are actually ascii files that are used in other contexts.
X Bitmap files are an example. They are formatted as 'C' style
#defines and an initialised array of characters, so that they can be included
in 'C' source code.  Adding trailing options would therefore render the file
unusable with a compiler, since it will get a syntax error on the railing option
string and the magic numbers. The solution to this is that xlito will ignore a
certain amount (a few hundred bytes) after the trailing options, and uses a
padding of 20 bytes before the trailing options. These two areas will be
maintained when changing an existing trailing option. In the case of an X
bitmap then, the solution is to edit the file and place the embedded string
in some 'C' comments:
.PP
eg: say the file starts as:
.br
#define tt_width 4
.br
#define tt_height 4
.br
static char tt_bits[] = {
.br
   0x08, 0x02, 0x04, 0x01};
.PP
and you add a trailing options:
.br
#define tt_width 4
.br
#define tt_height 4
.br
static char tt_bits[] = {
.br
   0x08, 0x02, 0x04, 0x01};
.br
01234567890123456789XXX xloadimage trailing options XXX0007"-smooth"0007XXX
.br
xloadimage trailing options XXX
.PP
Then the trailing options can be commented out:
.br
#define tt_width 4
.br
#define tt_height 4
.br
static char tt_bits[] = {
.br
   0x08, 0x02, 0x04, 0x01};
.br
/\(**234567890123456789XXX xloadimage trailing options XXX0007"-smooth"0007XXX
.br
xloadimage trailing options XXX \(**/
.SH BUGS
xlito doesn't cope with compressed files. A files will need uncompressing, the
options added, and then compressing again.




xlito.c
/* #ident	"@(#)x11:contrib/clients/xloadimage/xlito.c 1.8 94/07/29 Labtam" */

#ifndef NOT_STANDALONE
#include <fcntl.h>
#include <stdio.h>
#include <ctype.h>
#include <errno.h>
#endif /* NOT_STANDALONE */

#define VERSION "1"
#define PATCHLEVEL "02"

/* xli trailing option header file */

#define TOBUFSZ 1024
#define PADSIZE 20		/* allow a little over-read on image load */
#define KEY "XXX xloadimage trailing options XXX"		/* Magic number */

#ifndef NOT_STANDALONE	/* make header usable in xli, */
                        /* while keep xlito to one file */

#if (defined(SYSV) || defined (VMS)) && !defined(SVR4) && !defined(__hpux) && !defined(_CRAY)
#define NEED_FTRUNCATE		/* define this if your system doesn't have ftruncate() */
#endif

#define SHOW 0
#define XLI 1
#define CHANGE 2
#define DELETE 3

#ifdef MSDOS
#define EXTRA_FLAG O_BINARY
#else
#define EXTRA_FLAG 0
#endif

char *find_text();
char *pname, *fname;

#undef  DEBUG

#ifdef DEBUG
# define debug(xx)	fprintf(stderr,xx)
#else
# define debug(xx)
#endif


main(argc,argv)
int argc;
char *argv[];
{
	int func = SHOW;
	char *text;
	int a = 1;
	char *key = KEY;
	int klen;
	char *junk;
	int jlen;
	int file;
	int nameopt = 0;

	pname = argv[0];
	klen = strlen(key);
	if (argc < 2)
		usage(pname);

	if (argv[a][0] == '-')
	{
		if (argv[a][1] == 'c')
		{
			if (argv[a][2] != '\000')
				text = &argv[a][2];
			else if (argc >= 3)
				text = argv[++a];
			else
				usage(pname);
			func = CHANGE;
		}
		else if (argv[a][1] == 'd')
		{
			if (argv[a][2] != '\000')
				usage(pname);
			func = DELETE;
		}
		else if (argv[a][1] == 'x')
		{
			if (argv[a][2] != '\000')
				usage(pname);
			func = XLI;
		}
		else
			usage(pname);
		a++;
	}

    for(;a < argc; a++)	/* Do all the files */
	{
		fname = argv[a];
		if (func == XLI)
		{
			if (nameopt)
			{
				nameopt = 0;
				file = open(fname,O_RDONLY | EXTRA_FLAG);
				if (file != -1)
				{
					if ((text = find_text(file,key,&junk,&jlen, 1, 0)) != NULL)
						printf(" %s -name %s",text,fname);
					else
						printf(" -name %s",fname);
					close(file);
					continue;
				}
				printf(" -name %s",fname);
				continue;
			}
			if (argv[a][0] != '-')
			{
				file = open(fname,O_RDONLY | EXTRA_FLAG);
				if (file != -1)
				{
					if ((text = find_text(file,key,&junk,&jlen, 1, 0)) != NULL)
						printf(" %s %s",text,fname);
					else
						printf(" %s",fname);
					close(file);
					continue;
				}
			}
			else
			{
				if(strcmp("-name",fname)==0)
				{
					nameopt = 1;
					continue;
				}
			}
			printf(" %s",fname);
			continue;
		}
		else if (func == SHOW)
		{
			file = open(fname,O_RDONLY | EXTRA_FLAG);
			if (file != -1)
			{
				if ((text = find_text(file,key,&junk,&jlen, 1, 0)) != NULL)
				{
					printf("%s: '%s'\n",fname, text);
				}
				else
				{
					printf("%s: none\n",fname);
				}
				if(close(file) != 0)
				{
					fprintf(stderr,"%s: Error closing %s, ",pname,fname); perror("");
					exit(-1);
				}
				continue;
			}
			fprintf(stderr,"Unable to open file %s",fname);
			continue;
		}
		else if (func == DELETE)
		{
			file = open(fname,O_RDWR | EXTRA_FLAG);
			if (file == -1)
			{
				fprintf(stderr,"%s: Unable to open file '%s' for deleting options, ",pname,fname); perror("");
				exit(-1);
			}
			if ((text = find_text(file,key,&junk,&jlen, 0, 1)) != NULL)
			{
				long fsize;
				fsize = lseek(file, 0L,1);
				if(ftruncate(file, fsize) != 0)
				{
					fprintf(stderr,"%s: Error truncating %s, ",pname,fname); perror("");
					exit(-1);
				}
			}
			if(close(file) != 0)
			{
				fprintf(stderr,"%s: Error closing %s, ",pname,fname); perror("");
				exit(-1);
			}
			continue;
		}
		else if (func == CHANGE)
		{
			char padding[PADSIZE];
			int i;
			char tt[6];
			int tlen;
			long fsize;
			
			tlen = strlen(text);
			if (tlen > (TOBUFSZ - klen - 4))
			{
				fprintf(stderr,"%s: text is too long\n",pname);
				exit(-1);
			}
				
			file = open(fname,O_RDWR | EXTRA_FLAG);
			if (file == -1)
			{
				fprintf(stderr,"%s: Unable to open file '%s' for changing options, ",pname,fname); perror("");
				exit(-1);
			}
			if (!find_text(file,key,&junk,&jlen,0, 0))
			{	/* No padding there already, so add some */
				for(i=0; i < PADSIZE; i++) padding[i] = '0' + (i % 10);
				if (write(file, padding,PADSIZE) != PADSIZE)
				{
					fprintf(stderr,"%s: Error writing to file %s, ",pname,fname); perror("");
					exit(-1);
				}
			}
	
			sprintf(tt+1,"%04d",tlen);
			tt[0] = tt[ 5] = '"';
	
			if (write(file, key,klen) != klen
			|| write(file, tt + 1,5) != 5
			|| write(file, text,tlen) != tlen
			|| write(file, tt,5) != 5
			|| write(file, key,klen) != klen
			|| write(file, junk,jlen) != jlen)
			{
				fprintf(stderr,"%s: Error writing to file %s, ",pname,fname); perror("");
				exit(-1);
			}
			fsize = lseek(file, 0L,1);
			if(ftruncate(file, fsize) != 0)
			{
				fprintf(stderr,"%s: Error truncating %s, ",pname,fname); perror("");
				exit(-1);
			}
			if(close(file) != 0)
			{
				fprintf(stderr,"%s: Error closing %s, ",pname,fname); perror("");
				exit(-1);
			}
		}
	}
	return 0;
}


char buf1[TOBUFSZ];
char buf[TOBUFSZ];

char *
find_text(file,key,junk,jlen,show,delete)
int file;
char *key,**junk;
int *jlen;
int show;	/* Non zero if fail gracefully */
int delete;	/* Non zero to position at padding rather than at options */
{
	int klen;
	int count;
	int tlen;
	long foff, fsize, sk;
	char *pp, sc;
	
	klen = strlen(key);

	fsize = foff = lseek(file, 0L,2);

	/* Look for key backwards in file up to TOBUFSZ chars from end */
	sk = TOBUFSZ < fsize ? TOBUFSZ : fsize;
	if (lseek(file, -sk, 2) == -1L)
		goto seekerr;
	if ((count = read(file, buf1, sk)) != sk)
	{
		debug("didn't read key\n");
		goto fail;
	}
	sc = key[klen-1];
		
	for(pp = &buf1[sk-1];;pp--)
	{
		for(;*pp != sc && pp >= buf1; pp--);
		if(pp < (buf1 + klen -1))
		{
			debug("couldn't find key\n");
			goto fail;
		}
		if (strncmp(key,pp - klen + 1,klen) ==0)
			break;
	}
	/* found the last key string, so read the text length */
	*junk = pp + 1;
	*jlen = count + (buf1 - pp) - 1;
	sk = count + (buf1 - pp) - 1 + klen + 5;
	if ((foff = lseek(file, -(sk < fsize ? sk : fsize),2)) == -1L)
		goto seekerr;
	count = read(file, buf,5);
	if (count != 5)
	{
		debug("didn't read key\n");
		goto fail;
	}
	foff += count;
	if (buf[0] != '"' || !isdigit(buf[1]) || !isdigit(buf[2]) || !isdigit(buf[3]) || !isdigit(buf[4]))
	{
		debug("length wasn't recognised\n");
		goto fail;
	}
	buf[5] = '\000';
	tlen = atoi(&buf[1]);
	if (tlen > (TOBUFSZ - klen - 5))
	{
		debug("text was too long\n");
		goto fail;
	}
	sk = tlen + klen + 10;
	if ((foff = lseek(file, -(sk < foff ? sk : foff),1)) == -1L)
		goto seekerr;
	count = read(file, buf,(tlen + klen + 5));
	if (count != (tlen + klen + 5))
	{
		debug("didn't read key + text\n");
		goto fail;
	}
	foff += count;
	if (strncmp(key,buf,klen) !=0)
	{
		debug("first key wasn't recognised\n");
		goto fail;
	}
	if (!isdigit(buf[klen]) || !isdigit(buf[klen+1]) || !isdigit(buf[klen+2]) || !isdigit(buf[klen+3]) || buf[klen+4] != '"')
	{
		debug("first length wasn't recognised\n");
		goto fail;
	}
	buf[klen+4] = '\000';
	if (tlen != atoi(&buf[klen]))
	{
		debug("first length didn't match\n");
		goto fail;
	}
	sk = tlen + klen + 5;
	if (delete)
		sk += PADSIZE;
	if ((foff = lseek(file, -(sk < foff ? sk : foff),1)) == -1L)
		goto seekerr;
	buf[klen + 5 + tlen] = '\000';
	return &buf[klen + 5];

fail:
	*jlen = 0;
	if(lseek(file,0L,2) != -1L)
		return NULL;
seekerr:
	*jlen = 0;
	if (show)
		return NULL;
	fprintf(stderr,"%s: Error seeking on file %s, ",pname,fname); perror("");
	exit(-1);
}

usage(name)
char *name;
{
	fprintf(stderr,"%s: Version %s.%s usage\n",name,VERSION,PATCHLEVEL);
	fprintf(stderr,"%s [-x]|[-c text]|[-d] files\n",name);
	fprintf(stderr,"    -x          xli command line processing\n");
	fprintf(stderr,"    -c text     replace options with \"text\"\n");
	fprintf(stderr,"    -d          delete options\n");
	exit(-1);
}

#ifdef NEED_FTRUNCATE
#include <sys/types.h>
#include <unistd.h>

ftruncate(fd, size)
int fd;
off_t size;
{
	struct flock ff;
	
	ff.l_whence = SEEK_SET;
	ff.l_start = size;
	ff.l_len = 0;		/* till end of file */
	return(fcntl(fd, F_FREESP, (char *)&ff));
}

#endif

#endif /* NOT_STANDALONE */





